File: chapter03/python/subarraySum.py
# 时间复杂度为O(n^2), 空间复杂度为O(n)
def subarraySum(nums, k):
    sum_table = [0] * (len(nums) + 1)
    for i in range(len(nums)):
        sum_table[i+1] = sum_table[i] + nums[i]

    res = 0
    for i in range(1, len(nums)+1):
        for j in range(i):
            if sum_table[i] - sum_table[j] == k:
                res += 1
    return res


# 时间复杂度为O(n), 空间复杂度为O(n), 借助哈希表, 去除了不必要的嵌套
def subarraySumDict(nums, k):
    n = len(nums)
    d = {}
    d[0] = 1
    ans = 0
    sum0_i = 0

    for i in range(n):
        sum0_i += nums[i]
        sum0_j = sum0_i - k
        if sum0_j in d:
            ans += d.get(sum0_j)
        d[sum0_i] = d.get(sum0_i, 0) + 1
    return ans


if __name__ == "__main__":
    nums = [1, 1, 1]
    k = 2
    print(subarraySumDict(nums, k))

File: chapter03/cpp/subarraySum.cpp
#include <iostream>
#include <vector>
#include <string>
#include <algorithm>
#include <stdlib.h>
#include <unordered_map>

using namespace std;

int subarraySumDict(const vector<int>& nums, int target)
{
    unordered_map<int, int> m;
    int ans = 0;
    m[0] = 1;

    int sum_i = 0, sum_j;
    for(int i=0; i<nums.size(); i++)
    {
        sum_i += nums[i];
        sum_j = sum_i - target;
        m[sum_i] = m[sum_i] + 1;
        if(m.count(sum_j))
            ans += m[sum_j];
    }
    return ans;
}

int main()
{
    vector<int> nums = {3, 5, 2, -2, 4, 1};
    int target = 8;
    cout << subarraySumDict(nums, target) << endl;
    return 0;
}